### PR TITLE
Add forced actions flag

### LINKED ISSUE #117: Support forced actions when initial actions fail
Some actions fail because the target elements are overlayed by divs, but will work when setting the flag `force=True` ; by catching the initial exception and retrying with `force=True` inside the action call, we could spare some wasted actions from the agent.

### PR DESCRIPTION
**Pull request by jardinetsouffleton**
closes #117

### CODE REVIEW

File: browsergym/core/src/browsergym/core/action/functions.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 send_message_to_user: callable = None
 report_infeasible_instructions: callable = None
 demo_mode: Literal["off", "default", "all_blue", "only_visible_elements"] = None
+force_actions: bool = False
```
Comment (gasse) [inline browsergym/core/src/browsergym/core/action/functions.py:17]: I'd change the name for something more explicit, like `retry_with_force`. From `force_actions` one could think all actions are always executed with `force=True`.

File: browsergym/core/src/browsergym/core/action/functions.py [hunk 2]
```diff
@@ -82,8 +88,15 @@ def check(bid: str):
         check('55')
     """
     elem = get_elem_by_bid(page, bid, demo_mode != "off")
-    add_demo_mode_effects(page, elem, bid, demo_mode=demo_mode, move_cursor=True)
-    elem.check(timeout=500)
+    if demo_mode != "off":
```
Comment (gasse) [inline browsergym/core/src/browsergym/core/action/functions.py:91]: Not needed, a check is done inside the function
https://github.com/ServiceNow/BrowserGym/blob/b21860d0df4a3d5396ac5afcd5cfdf61bd9d341a/browsergym/core/src/browsergym/core/action/utils.py#L244

File: browsergym/core/src/browsergym/core/action/functions.py [hunk 3]
```diff
@@ -82,8 +88,15 @@ def check(bid: str):
...
+        add_demo_mode_effects(page, elem, bid, demo_mode=demo_mode, move_cursor=True)
+    if force_actions:
+        try:
+            elem.check(timeout=500)
+        except Exception as e:
+            elem.click(force=True, timeout=500)
```
Comment (gasse) [inline browsergym/core/src/browsergym/core/action/functions.py:97]: Same here, check takes a force argument
```suggestion
            elem.check(timeout=500, force=True)
```

File: browsergym/core/src/browsergym/core/action/functions.py [hunk 4]
```diff
@@ -95,8 +108,15 @@ def uncheck(bid: str):
...
-    elem.uncheck(timeout=500)
+    if demo_mode != "off":
+        add_demo_mode_effects(page, elem, bid, demo_mode=demo_mode, move_cursor=True)
+    if force_actions:
+        try:
+            elem.uncheck(timeout=500)
+        except Exception as e:
+            elem.click(force=True, timeout=500)
+    else:
+        elem.uncheck(timeout=500)
```
Comment (gasse) [inline browsergym/core/src/browsergym/core/action/functions.py:119]: Same comments

File: browsergym/core/src/browsergym/core/action/functions.py [hunk 5]
```diff
@@ -110,8 +130,15 @@ def select_option(bid: str, options: str | list[str]):
         select_option('c48', ["red", "green", "blue"])
     """
     elem = get_elem_by_bid(page, bid, demo_mode != "off")
-    add_demo_mode_effects(page, elem, bid, demo_mode=demo_mode, move_cursor=False)
-    elem.select_option(options, timeout=500)
+    if demo_mode != "off":
```
Comment (gasse) [inline browsergym/core/src/browsergym/core/action/functions.py:133]: same, not needed

File: browsergym/core/src/browsergym/core/action/functions.py [hunk 6]
```diff
@@ -129,8 +156,15 @@ def click(
         click('48', button="middle", modifiers=["Shift"])
     """
     elem = get_elem_by_bid(page, bid, demo_mode != "off")
-    add_demo_mode_effects(page, elem, bid, demo_mode=demo_mode, move_cursor=True)
-    elem.click(button=button, modifiers=modifiers, timeout=500)
+    if demo_mode != "off":
```
Comment (gasse) [inline browsergym/core/src/browsergym/core/action/functions.py:159]: same, not needed

File: browsergym/core/src/browsergym/core/action/functions.py [hunk 7]
```diff
@@ -148,8 +182,15 @@ def dblclick(
         dblclick('178', button="middle", modifiers=["Shift"])
     """
     elem = get_elem_by_bid(page, bid, demo_mode != "off")
-    add_demo_mode_effects(page, elem, bid, demo_mode=demo_mode, move_cursor=True)
-    elem.dblclick(button=button, modifiers=modifiers, timeout=500)
+    if demo_mode != "off":
```
Comment (gasse) [inline browsergym/core/src/browsergym/core/action/functions.py:185]: same, not needed

File: tests/core/test_actions_highlevel.py [hunk 8]
```diff
@@ -1399,3 +1400,42 @@ def get_checkbox_elem(obs):
     # box not checked
     assert not obs["last_action_error"]
     assert not checkbox.has_attr("checked")
+
+
+# test that forced action can click an obstructed element
+@pytest.mark.parametrize("force_actions", [True, False])
+def test_forced_actions(force_actions):
```
Comment (gasse) [inline tests/core/test_actions_highlevel.py:1407]: Simple et efficace, love this test :)

**Review by gasse — CHANGES_REQUESTED** [review]

File: tests/core/test_actions_highlevel.py (hunk 8 above)
Comment (jardinetsouffleton) [reply tests/core/test_actions_highlevel.py:1407]: :)

File: browsergym/core/src/browsergym/core/action/functions.py (hunk 1 above)
Comment (jardinetsouffleton) [reply browsergym/core/src/browsergym/core/action/functions.py:17]: agreed

File: browsergym/core/src/browsergym/core/action/functions.py (hunk 2 above)
Comment (jardinetsouffleton) [reply browsergym/core/src/browsergym/core/action/functions.py:91]: oh! ok will revert

**Review by gasse — APPROVED** [review]
LGTM

**Review by gasse — APPROVED** [review]

**Pull request merged**